addons/functions/functions/fnc_hitEH.sqf

```diff
@@ -0,0 +1,60 @@
+#include "script_component.hpp"
+/*
+ * Author: nomisum
+ * Receiving and calculating Damage
+ *
+ * Arguments:
+ * 0: Trench <OBJECT>
+ * 1: Multiplier <NUMBER>
+ *
+ * Return Value:
+ * None
+ *
+ * Example:
+ * [TrenchObj, 1] call grad_trenches_functions_fnc_hitEH;
+ *
+ * Public: No
+ */
+
+params ["_trench", "_multiplier"];
+
+if (_trench getVariable [QGVAR(hitHandler), false]) exitWith { 
+    TRACE_1("hitEH","not adding hit handler, already there");
+};
+
+_trench setVariable [QGVAR(hitHandler), true, true];
+
+
+
+_trench addEventHandler ["HitPart", {
+    (_this select 0) params ["_trench", "_shooter", "_projectile", "_position", "_velocity", "_selection", "_ammo", "_vector", "_radius", "_surfaceType", "_isDirect"];
+
+    diag_log str _isDirect;
+
+    diag_log str (_ammo select 2);
+
+    if (!_isDirect) then {
+        // add cooldown after indirect
+        _ammo params ["", "", "_splashDamage", "", "_type"];
+
+        // diag_log format ["direct Hit with %1", _ammo];
+
+        if (_splashDamage > 1) then {
+            // send fx to clients
+            [vectorDir _trench, ASLToAGL _position] remoteExec [QFUNC(hitFX), [0,-2] select isDedicated];
+
+            private _progress = _trench getVariable ["ace_trenches_progress", 0];
+            private _multiplier = [configFile >> "CfgVehicles" >> typeOf _trench >> "grad_trenches_damageMultiplier", "NUMBER", 1] call CBA_fnc_getConfigEntry; 
+            private _damage = (_splashDamage*_multiplier)/20; // 1 HE shell appr 25% decay depending on ammo type
+            _progress = _progress - _damage;
+
+            if (_progress > 0) then {
+                private _lift = linearConversion [0, 1, _progress, -([configFile >> "CfgVehicles" >> typeOf _trench >> QGVAR(offset), "NUMBER", 2] call CBA_fnc_getConfigEntry), 0, true];
+                _trench animateSource ["rise", _lift, true];
+                _trench setVariable ["ace_trenches_progress", _progress, true];
```

Comment:
I smell a race condition here.

Imagine I'm digging a trench. The trench is local to my client (I guess? Or is it the server?). Every 100ms the trench is raised by a bit.

Now you, a different client, shoot up the trench. You're setting the trench progress here (okay, this handler runs on the server only, but – see above reference to the HitPart docs – we'll probably need to add it on every client).

* My client – in the digging PFH – gets the current progress (lets say 0.7)
* Your client's "set progress to 0.4 because I shot it" value arrives
* My client – in the digging PFH – sets the progress it got plus the new progress (0.7+0.05 or so)
-> your damage goes unnoticed; maybe the trench yanks down and up again)

Reply:
Suggestion:
animateSource and ace_trenches_progress should ideally be *set AND read* on one node only (client or server, whatever), ideally where the trench is local.

Everything else should work via events, and the node where the trench is local can then make a good decision what to do about them and keep things consistent

Reply:
yes, locality should be fixed for the trench after placing. its not necessary to have fluent building anymore with animations i hope. before it was not possible due to significant lags of setpos.

Reply:
Locality is changed now, might still need some clean up